src/odemis/driver/actuator.py

```diff
@@ -716,53 +727,160 @@ def _updateReferenced(self, dep_refd):
+        # pos_val is a vector, so relative conversion
+        shift_dep = self._get_pos_vector(shift, absolute=False)
         logging.debug("converted relative move from %s to %s", shift, shift_dep)
-        f = self._dependency.moveRel(shift_dep, **kwargs)
-        return f
+        return self._dependency.moveRel(shift_dep, **kwargs)
 
     @isasync
     def moveAbs(self, pos, **kwargs):
         """
         **kwargs: Mostly there to support "update" argument
         """
         # pos is a position, so absolute conversion
-        cpos = self.position.value
-        vpos = pos.get("x", cpos["x"]), pos.get("y", cpos["y"])
-        vpos_dep = self._convertPosTodep(vpos)
-
-        pos_dep = {self._axes_dep["x"]: vpos_dep[0],
-                     self._axes_dep["y"]: vpos_dep[1]}
+        pos_dep = self._get_pos_vector(pos)
         logging.debug("converted absolute move from %s to %s", pos, pos_dep)
-        f = self._dependency.moveAbs(pos_dep, **kwargs)
-        return f
+        return self._dependency.moveAbs(pos_dep, **kwargs)
 
     def stop(self, axes=None):
         self._dependency.stop()
 
     @isasync
     def reference(self, axes):
-        dep_axes = set(self._axes_dep[a] for a in axes)
-        f = self._dependency.reference(dep_axes)
-        return f
+        dep_axes = {self._axes_dep[a] for a in axes}
+        return self._dependency.reference(dep_axes)
 
+class Convert3DStage(ConvertStage):
+    """
+    Extends original ConvertStage with an additional axis Z
+    """
+    def __init__(self, name, role, dependencies, axes,
+                 rotation=(0, 0, 0), scale=(1, 1, 1), translation=(0, 0, 0), **kwargs):
+        """
+        dependencies (dict str -> actuator): name to objective lens actuator
+        axes (list of 3 strings): names of the axes for x, y and z
+        scale (None tuple of 3 floats): scale factor from exported to original position
+        rotation (tuple of 3 floats): rotation factor rz, ry, rx (in radians)
+        translation (None or tuple of 3 floats): translation offset (in m)
+        """
+        assert len(axes) == 3
+        if len(dependencies) != 1:
+            raise ValueError("ConvertStage needs 1 dependency")
+
+        self._dependency = list(dependencies.values())[0]
+        self._axes_dep = {"x": axes[0], "y": axes[1], "z": axes[2]}
+        if rotation.count(0) < 2:
+            raise ValueError("Convert3DStage allows only one rotation angle to be > 0.")
+
+        axes_def = {"x": self._dependency.axes[axes[0]],
+                    "y": self._dependency.axes[axes[1]],
+                    "z": self._dependency.axes[axes[2]]}
+        model.Actuator.__init__(self, name, role, dependencies=dependencies, axes=axes_def, **kwargs)
+
+        self._metadata[model.MD_POS_COR] = translation
+        self._metadata[model.MD_ROTATION_COR] = rotation
+        self._metadata[model.MD_PIXEL_SIZE_COR] = scale
+        self._updateConversion()
```

Comment:
Aren't (or shouldn't) MD_ROTATION_COR and MD_PIXEL_SIZE_COR be properties of the imaging system instead of the stage?

Reply:
I think the reason of using them in the stage is just reuse the metadata that has `correction` in its name, to not recreate new metadata, is that correct @pieleric ?

Reply:
We reuse these metadata (and the metadata API in general) as a way to reconfigure dynamically the wrapper. It's actually not something introduced by Bassim, it was already there for a long while...
Not that typically the values are not changed frequently, it's just during calibration, at system install, or during alignment.